### PR TITLE
Add cmesh examples triangulated spherical shell (octahedron & icosahedron versions) [5/n]. 

### PR DESCRIPTION
**Pull request by jmark**
**_Describe your changes here:_**

This is a follow-up to https://github.com/DLR-AMR/t8code/pull/791.
grafik

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

#### General
- [x] The reviewer executed the new code features at least once and checked the results manually

- [x] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [x] New source/header files are properly added to the Makefiles
- [x] The code is well documented
- [x] All function declarations, structs/classes and their members have a proper doxygen documentation
- [x] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)

#### Tests
- [ ] The code is covered in an existing or new test case using Google Test

#### Github action

- [x] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)
- [x] All tests pass (in various configurations, this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

#### Scripts and Wiki

- [ ] If a new directory with source-files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [ ] If this PR introduces a new feature, it must be covered in an example/tutorial and a Wiki article.

#### Licence

- [x] The author added a BSD statement to `doc/` (or already has one)

### CODE REVIEW

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.hxx [hunk 1]
```diff
@@ -99,17 +99,50 @@ struct t8_geometry_triangulated_spherical_surface: public t8_geometry_with_verti
   }
 
   /**
-   * Map the faces of an oktaeder to a spherical surface.
+   * Map the faces of an octahedron/icasohedron to a spherical surface.
```
Comment (lukasdreyer) [inline src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.hxx:102]: Please do a search and replace for `icasohedron` and `icosahedron`
```suggestion
   * Map the faces of an octahedron/icosahedron to a spherical surface.
```

**Review by lukasdreyer — CHANGES_REQUESTED** [review]
First look, I have not yet looked at the cmesh construction in detail.

File: src/t8_cmesh/t8_cmesh_examples.c [hunk 2]
```diff
@@ -2868,6 +2868,138 @@ t8_cmesh_new_triangulated_spherical_surface (const double radius, sc_MPI_Comm co
...
+  double all_verts[ntrees * T8_ECLASS_MAX_CORNERS * T8_ECLASS_MAX_DIM];
+  t8_eclass_t all_eclasses[ntrees];
+
+  /* Defitition of the tree class. */
+  for (int itree = 0; itree < ntrees; itree++) {
+    t8_cmesh_set_tree_class (cmesh, itree, T8_ECLASS_TRIANGLE);
+    all_eclasses[itree] = T8_ECLASS_TRIANGLE;
+  }
+
+  const double alpha = 63.43494882292201 / 180.0 * M_PI; /* Icasohedral angle. */
```
[the remark this answers is not in the document]
Comment (jmark) [reply src/t8_cmesh/t8_cmesh_examples.c:2895]: Done.

File: src/t8_cmesh/t8_cmesh_examples.c [hunk 3]
```diff
@@ -2868,6 +2868,138 @@ t8_cmesh_new_triangulated_spherical_surface (const double radius, sc_MPI_Comm co
...
+
+  double vertices_top[3 * 3];
+  double vertices_bot[3 * 3];
+
+  {
+    /* Prepare initial triangle on top top of the icosahedron. */
```
[the remark this answers is not in the document]
Comment (jmark) [reply src/t8_cmesh/t8_cmesh_examples.c:2901]: Done.

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx [hunk 4]
```diff
@@ -266,28 +280,87 @@ t8_geometry_triangulated_spherical_surface::t8_geom_evaluate (t8_cmesh_t cmesh,
...
       const double ww = u_ref[1] + x * v_ref[1] + y * w_ref[1];
 
-      /* Correction in order to rectify elements near the corners. */
+      /* tldr: Correction in order to rectify elements near the corners. This
+       * is necessary, since due to the transformation from the cmesh triangle to the
+       * sphere elements near the face centers expand while near the corners they
+       * shrink. Following correction alleviates this.
+       * TODO: This correction is not general and not optimal in all cases.
+       *       Find a better one. (This is not a trivial task, though.)
+       */
```
[the remark this answers is not in the document]
Comment (jmark) [reply src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx:289]: Yes!

Refactored this code. Should be clearer now.

File: src/t8_cmesh/t8_cmesh_examples.c [hunk 5]
```diff
@@ -2868,6 +2868,138 @@ t8_cmesh_new_triangulated_spherical_surface (const double radius, sc_MPI_Comm co
...
+{
+  /* Initialization of the mesh */
+  t8_cmesh_t cmesh;
+  t8_cmesh_init (&cmesh);
+
+  t8_geometry_c *geometry = t8_geometry_triangulated_spherical_surface_new ();
+
+  t8_cmesh_register_geometry (cmesh, geometry); /* Use linear geometry */
+
+  const int ntrees = 20; /* Number of cmesh elements resp. trees. */
```
[the remark this answers is not in the document]
Comment (jmark) [reply src/t8_cmesh/t8_cmesh_examples.c:2882]: Done.

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx [hunk 6]
```diff
@@ -110,28 +110,28 @@ t8_geometry_squared_disk::t8_geom_evaluate (t8_cmesh_t cmesh, t8_gloidx_t gtreei
...
+  r[2] = active_tree_vertices[2];
+  t8_vec_normalize (r);
 
   /* The next three code blocks straighten out the elements near the triangle
    * corners by averaging the rectification with all three corners. */
 
-  /* First triangle corner. */
+  /* With this factor we compute the intersection of `r` and `p` (see further
+   * below) and average over the three corners of the triangle. */
+  const double factor = 1.0 / (r[0] * n[0] + r[1] * n[1] + r[2] * n[2]) / 3.0;
```
[the remark this answers is not in the document]
Comment (jmark) [reply src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx:132]: Nice tip! Didn't think of `t8_vec_dot`!

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx [hunk 7]
```diff
@@ -110,28 +110,28 @@ t8_geometry_squared_disk::t8_geom_evaluate (t8_cmesh_t cmesh, t8_gloidx_t gtreei
...
+
+  /* First triangle/prism corner. */
```
[the remark this answers is not in the document]
Comment (jmark) [reply src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx:134]: Done.

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.hxx (hunk 1 above)
Comment (jmark) [reply src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.hxx:102]: Done.

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx [hunk 8]
```diff
@@ -110,184 +110,169 @@ t8_geometry_squared_disk::t8_geom_evaluate (t8_cmesh_t cmesh, t8_gloidx_t gtreei
...
- */
-void
-t8_geometry_triangulated_spherical_surface::t8_geom_evaluate (t8_cmesh_t cmesh, t8_gloidx_t gtreeid,
-                                                              const double *ref_coords, const size_t num_coords,
-                                                              double *out_coords) const
+/* Helper function for `t8_geom_evaluate_sphere_tri_prism`. */
+static inline void
+t8_map_triangle_to_sphere (const double *active_tree_vertices, const double sphere_radius, const int shift,
+                           const double u_ref[3], const double v_ref[3], const double w_ref[3],
+                           const double *ref_coords, const size_t num_coords, double *out_coords)
```
[the remark this answers is not in the document]
Comment (jmark) [reply src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx:117]: I moved the 1/3 as you suggested. Done.

File: src/t8_cmesh/t8_cmesh_examples.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (jmark) [reply src/t8_cmesh/t8_cmesh_examples.c]: Fixed. :)

**Review by lukasdreyer — APPROVED** [review]

**Pull request merged**